### PR TITLE
[exo-deploy - 1] Generate aws module

### PR DESCRIPTION
**Pull request by hugoesthere**
First iteration of exo-deploy, it just generates a couple modules containing application config. Working on generating service modules next.

### CODE REVIEW

File: exo-go/bin/setup [hunk 1]
```diff
@@ -1,4 +1,6 @@
 #!/usr/bin/env bash
 
 go get -u gopkg.in/alecthomas/gometalinter.v1
+go get -u github.com/jteeuwen/go-bindata
```
Comment (charlierudolph) [inline exo-go/bin/setup]: this can be combined with the previous line
```
go get -u gopkg.in/alecthomas/gometalinter.v1 github.com/jteeuwen/go-bindata
```

File: exo-go/lint-config.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+  "Exclude": ["src/terraform_helpers/bindata.go"]
+}
```
Comment (charlierudolph) [inline exo-go/lint-config.json]: 👍

File: exo-go/src/terraform_file_helpers/template_builder.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	"github.com/hoisie/mustache"
+)
+
+// RenderTemplates renders a Terraform template
+func RenderTemplates(templateName string, varsMap map[string]string) string {
+	template := getTemplate(templateName)
+	return mustache.Render(template, varsMap)
+}
+
+// getTemplate returns a stringified template
```
Comment (charlierudolph) [inline exo-go/src/terraform_file_helpers/template_builder.go]: Comments are only required on public functions.

File: exo-go/src/terraform_file_helpers/template_builder.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+	var filePerm os.FileMode = 0744 //standard Unix file permission: rwxrw-rw-
+	err = os.MkdirAll(filepath.Join(cwd, "terraform"), filePerm)
+	if err != nil {
+		log.Fatalf("Failed to get create directory: %s", err)
+	}
+
+	err = ioutil.WriteFile(filepath.Join(cwd, "terraform", "main.tf"), []byte(data), filePerm)
+	if err != nil {
+		log.Fatalf("Failed writing Terraform files: %s", err)
```
Comment (charlierudolph) [inline exo-go/src/terraform_file_helpers/template_builder.go]: Please no log.Fatal / panic in helper functions. Please only call this from the top level command. Instead this function should return an error. We can wrap the errors via https://github.com/pkg/errors to give extra context.

File: .travis.yml [hunk 5]
```diff
@@ -29,7 +29,8 @@ before_install:
 
 
 install:
-  - yarn install
+  - mkdir node_modules/
+  - mkdir go-binaries/
```
Comment (hugoesthere) [inline .travis.yml:33]: This is an attempt to work around Morula. `yarn install` is no longer needed on the root level, and is causing issues because it's running the pre-publish script `bin/cross-compile` before `./morula all bin/setup` is run.  `bin/cross-compile` uses a command `go-bindata`, which is installed in the `exo-go/bin/setup` script. We can simply run `./morula all bin/setup`, but it throws an error when ignored directories haven't been created yet (`node_modules/` and `go-binaries/` in this case).

We're working on a PR for Morula so it doesn't throw when trying to ignore non-existent directories.

**Comment by hugoesthere** [comment]
@charlierudolph ready for another look.

File: bin/cross-compile [hunk 6]
```diff
@@ -1,6 +1,10 @@
 #!/usr/bin/env bash
 set -ex
 
+cd exo-go
```
Comment (kevgo) [inline bin/cross-compile:4]: Why are you moving test infrastructure around in this PR that is to create exo-deploy? Can you please extract this separate change into its own PR?

File: exo-go/lint-config.json [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
+  "Exclude": ["src/terraform_file_helpers/bindata.go"]
```
Comment (kevgo) [inline exo-go/lint-config.json:2]: Cool!

File: exo-go/features/help.feature [hunk 8]
```diff
@@ -54,6 +55,15 @@ Feature: help command
         exo create
       """
 
+  Scenario: the user enters 'exo deploy help'
+    When running "exo deploy help" in the terminal
```
Comment (kevgo) [inline exo-go/features/help.feature]: The command to see help is `exo help deploy`

File: exo-go/src/cmd/deploy.go [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/Originate/exosphere/exo-go/src/terraform_file_helpers"
+	"github.com/spf13/cobra"
+)
+
+// deployCmd represents the deploy command
+var deployCmd = &cobra.Command{
+	Use:   "deploy",
+	Short: "Deploys Exosphere application to the cloud",
+	Run: func(cmd *cobra.Command, args []string) {
+		if printHelpIfNecessary(cmd, args) {
```
Comment (kevgo) [inline exo-go/src/cmd/deploy.go]: What is this method for? Cobra has built-in functionality for showing help. You have to define a `Long` entry in the `cobra.Command` struct for that.

File: exo-go/src/cmd/deploy.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+			return
+		}
+		fmt.Println("We are about to deploy an application!")
+
+		appConfig, err := appConfigHelpers.GetAppConfig()
+		if err != nil {
+			log.Fatalf("Failed to read application configuration: %s", err)
```
Comment (kevgo) [inline exo-go/src/cmd/deploy.go]: Shorter and more active error message: `Cannot read application configuration: ...`

File: exo-go/src/terraform_file_helpers/index.go [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/pkg/errors"
+)
+
+// GenerateTerraform generates the main terraform file given application and service configuration
+func GenerateTerraform(appConfig types.AppConfig) error {
+	fileData := []string{}
+
+	moduleData, err := generateAwsModule(appConfig)
+	if err != nil {
+		return errors.Wrap(err, "Failed to generate Terraform")
```
Comment (kevgo) [inline exo-go/src/terraform_file_helpers/index.go]: Failed to generate __AWS__ Terraform module

**Review by kevgo — COMMENTED** [review]
This is looking promising! Good idea to bundle the Terraform templates into the binary using go-bindata!

File: exo-go/src/cmd/deploy.go (hunk 9 above)
Comment (hugoesthere) [reply exo-go/src/cmd/deploy.go]: Yes, it calls a shared method that every other command has been using. Looks like that method leverages the cobra `Help` function.

File: bin/cross-compile (hunk 6 above)
Comment (hugoesthere) [reply bin/cross-compile:4]: This goes a couple layers deep. `bin/cross-compile` needs to be run at the root level so the `go-binaries/` directory exists on the root level as well. This used to be done in the `pre-publish` script defined in `package.json`, but since we're no longer running `yarn install` at the root during setup (per my comment above), we have to move `cross-compile` to the root level and run it there ourselves.

File: exo-go/src/terraform_file_helpers/template_builder.go [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+	template, err := getTemplate(templateName)
+	if err != nil {
+		return "", err
+	}
+	return mustache.Render(template, varsMap), nil
+}
+
+// WriteTerraformFile writes the main Terraform file to the path: cwd/terraform/main.tf
+func WriteTerraformFile(data string) error {
+	cwd, err := os.Getwd()
```
Comment (charlierudolph) [inline exo-go/src/terraform_file_helpers/template_builder.go]: Please pass in the cwd / appDir in as a parameter. Lets just get the working directory once in the command and pass it to the helper functions as needed. This also makes things easier to test.

**Review by charlierudolph — APPROVED** [review]
